firedrake/slate/slac/kernel_builder.py

```diff
@@ -653,54 +745,247 @@ def slate_call(self, prg, temporaries):
+        str2name.update({"A_on_x": A_on_x_name, "A_on_p": A_on_p_name})
+
+        # setup the stop criterions
+        str2name.update({"stop_criterion_id": "cond", "stop_criterion_dep": "rkp1_normk"})
+        stop_criterion = self.generate_code_for_stop_criterion("rkp1_norm",
+                                                               1.e-16,
+                                                               str2name["stop_criterion_dep"],
+                                                               str2name["stop_criterion_id"])
+        str2name.update({"preconverged_criterion_id": "projis0", "preconverged_criterion_dep": "projector"})
+        corner_case = self.generate_code_for_converged_pre_iteration(str2name["preconverged_criterion_dep"],
+                                                                     str2name["preconverged_criterion_id"])
+
+        # NOTE The last line in the loop to convergence is another WORKAROUND
+        # bc the initialisation of A_on_p in the action call does not get inlined properly either
+        knl = loopy.make_function(
+            """{[i_0,i_1,i_2,i_3,i_4,i_5,i_6,i_7,i_8,i_9,i_10,i_11,i_12]:
+                 0<=i_0,i_1,i_2,i_3,i_4,i_5,i_7,i_8,i_9,i_10,i_11,i_12<n
+                 and 0<=i_6<=n}""",
+            ["""{x}[i_0] = -{b}[i_0] {{id=x0}}
+                {A_on_x}[:] = action_A({A}[:,:], {x}[:]) {{dep=x0, id=Aonx}}
+                <> r[i_3] = {A_on_x}[i_3]-{b}[i_3] {{dep=Aonx, id=residual0}}
+                {p}[i_4] = -r[i_4] {{dep=residual0, id=projector0}}
+                <> rk_norm = 0. {{dep=projector0, id=rk_norm0}}
+                rk_norm = rk_norm + r[i_5]*r[i_5] {{dep=projector0, id=rk_norm1}}
+                for i_6
+                    {A_on_p}[:] = action_A_on_p({A}[:,:], {p}[:]) {{dep=Aonp0, id=Aonp, inames=i_6}}
+                    <> p_on_Ap = 0. {{dep=Aonp, id=ponAp0}}
+                    p_on_Ap = p_on_Ap + {p}[i_2]*{A_on_p}[i_2] {{dep=ponAp0, id=ponAp}}
+                    <> projector_is_zero = abs(p_on_Ap) < 1.e-16 {{id={preconverged_criterion_dep}, dep=ponAp}}
+             """.format(**str2name),
+             corner_case,
+             """    <> alpha = rk_norm / p_on_Ap {{dep={preconverged_criterion_id}, id=alpha}}
+                    {x}[i_7] = {x}[i_7] + alpha*{p}[i_7] {{dep=ponAp, id=xk}}
+                    r[i_8] = r[i_8] + alpha*{A_on_p}[i_8] {{dep=xk,id=rk}}
+                    <> rkp1_norm = 0. {{dep=rk, id=rkp1_norm0}}
+                    rkp1_norm = rkp1_norm + r[i_9]*r[i_9] {{dep=rkp1_norm0, id={stop_criterion_dep}}}
+             """.format(**str2name),
+             stop_criterion,
+             """    <> beta = rkp1_norm / rk_norm {{dep={stop_criterion_id}, id=beta}}
+                    rk_norm = rkp1_norm {{dep=beta, id=rk_normk}}
+                    {p}[i_10] = beta * {p}[i_10] - r[i_10] {{dep=rk_normk, id=projectork}}
+                    {A_on_p}[i_12] = 0. {{dep=projectork, id=Aonp0, inames=i_6}}
+                end
```

Comment:
I imagine that this bit needs extensive commenting.

Reply:
And potentially an f-string.

Reply:
Okay I thought about the commenting. Obviously all of this looks a bit crazy because the kernel is implemented in a string. But really this is just the conjugate gradient algorithm where all matrix-vector products are replaced through actions. So I added this in the function docs, but I think I don't need to add more.